plugins/com.google.cloud.tools.eclipse.appengine.login.test/src/com/google/cloud/tools/eclipse/appengine/login/GoogleLoginServiceTest.java

```diff
@@ -33,64 +40,171 @@
+    Account account = loginService.logIn(null);
+
+    assertNull(account);
+    assertFalse(loginService.hasAccounts());
+    assertTrue(loginService.getAccounts().isEmpty());
+  }
+
+  @Test
+  public void testMultipleLogins() {
+    GoogleLoginService loginService = newLoginServiceWithMockLoginState(true);
+
+    loginService.logIn(null /* no dialog message */);
+    Set<Account> accounts1 = loginService.getAccounts();
+    assertEquals(1, accounts1.size());
+    assertTrue(accounts1.contains(account1));
+
+    loginService.logIn(null);
+    Set<Account> accounts2 = loginService.getAccounts();
+    assertEquals(2, accounts2.size());
+    assertTrue(accounts2.contains(account1));
+    assertTrue(accounts2.contains(account2));
+
+    loginService.logIn(null);
+    Set<Account> accounts3 = loginService.getAccounts();
+    assertEquals(3, accounts3.size());
+    assertTrue(accounts3.contains(account1));
+    assertTrue(accounts3.contains(account2));
+    assertTrue(accounts3.contains(account3));
+  }
+
+  @Test
+  public void testLogOutAll() {
+    GoogleLoginService loginService = newLoginServiceWithMockLoginState(true);
+
+    loginService.logIn(null /* no dialog message */);
+    loginService.logIn(null);
+    loginService.logIn(null);
+
+    assertTrue(loginService.hasAccounts());
+    assertFalse(loginService.getAccounts().isEmpty());
+
+    loginService.logOutAll();
+
+    assertFalse(loginService.hasAccounts());
+    assertTrue(loginService.getAccounts().isEmpty());
+  }
+
+  @Test
+  public void testGoogleLoginService_removeSavedCredentialIfNullRefreshToken() {
+    when(savedOAuthData.getEmail()).thenReturn("my-email@example.com");
+    when(savedOAuthData.getStoredScopes()).thenReturn(OAUTH_SCOPES);
+    when(savedOAuthData.getRefreshToken()).thenReturn(null);
+
+    new GoogleLoginService(dataStore, uiFacade, loggerFacade);
+    verify(dataStore, times(1)).removeOAuthData("my-email@example.com");
+  }
+
+  @Test
+  public void testGoogleLoginService_removeSavedCredentialIfScopesChanged() {
     // Persisted credential in the data store has an out-dated scopes.
-    SortedSet<String> newScope = new TreeSet<String>(Arrays.asList("new scope"));
+    Set<String> newScope = new HashSet<>(Arrays.asList("new_scope"));
+    when(savedOAuthData.getEmail()).thenReturn("my-email@example.com");
     when(savedOAuthData.getStoredScopes()).thenReturn(newScope);
     when(savedOAuthData.getRefreshToken()).thenReturn("fake_refresh_token");
 
-    GoogleLoginService loginService = new GoogleLoginService(dataStore, uiFacade, loggerFacade);
-    Assert.assertNull(loginService.getCachedActiveCredential());
+    new GoogleLoginService(dataStore, uiFacade, loggerFacade);
+    verify(dataStore, times(1)).removeOAuthData("my-email@example.com");
   }
 
   @Test
   public void testGoogleLoginService_restoreSavedCredential() {
     // Persisted credential in the data store is valid.
```

Comment:
Nit: "persisted" is redundant with "in the data store". Suggest just "Credential in the data store is valid"